clarify behavior of topo constraints (#28495)
kubernetes/website: content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md

You can define one or multiple `topologySpreadConstraint` to instruct the kube-scheduler how to place each incoming Pod in relation to the existing Pods across your cluster. The fields are:

- **maxSkew** describes the degree to which Pods may be unevenly distributed.
  It must be greater than zero. Its semantics differs according to the value of `whenUnsatisfiable`:
  - when `whenUnsatisfiable` equals to "DoNotSchedule", `maxSkew` is the maximum
    permitted difference between the number of matching pods in the target
    topology and the global minimum 

>     (global minimum is the minimum number of pods matching the label selector in all topology domains).

Reviewer: this is still not reading well. Any suggestions @sftim ?

Example: you have 3 zones with 0, 2 and 3 matching pods respectively. The global minimum is 0.

Author: It helps to understand by adding an example right after the global minimum definition. How about we change it to
```
(global minimum is the minimum number of pods matching the label selector in all topology domains. For example, if you have 3 zones with 0, 2 and 3 matching pods respectively, The global minimum is 0.)
```

Reviewer: I think we don't need to repeat "global minimum" within the parenthesis.

Other than that, I think a better phrasing is `(the minimum number of pods that match the label selector in a topology domain. For example...`

  zero. Its semantics differs according to the value of `whenUnsatisfiable`:
  - when `whenUnsatisfiable` equals to "DoNotSchedule", `maxSkew` is the maximum
    permitted difference between the number of matching pods in the target

>     topology and the global minimum (global minimum is the minimum number of pods in all topology domains).

Reviewer 2: > global minimum is the minimum number of pods in all topology domains

Is this global minimum a singleton value at the cluster level, or is it “global” because it's across multiple topology domains but also specific to the label selector being used?

(I assume the latter, but let's make it clear)

Author: You are right that it's the latter. Updated the doc.

  zero. Its semantics differs according to the value of `whenUnsatisfiable`:
  - when `whenUnsatisfiable` equals to "DoNotSchedule", `maxSkew` is the maximum
    permitted difference between the number of matching pods in the target
    topology and the global minimum 

>     (global minimum is the minimum number of pods in all topology domains that match the label selector).

Reviewer 3: Yeah I actually wanted to draw a diagram to show what global minimum means, but then I stopped shot as I did not know if it were possible to upload diagrams here.

  - `ScheduleAnyway` tells the scheduler to still schedule it while prioritizing nodes that minimize the skew.
- **labelSelector** is used to find matching Pods. Pods that match this label selector are counted to determine the number of Pods in their corresponding topology domain. See [Label Selectors](/docs/concepts/overview/working-with-objects/labels/#label-selectors) for more details.

> When a Pod defines more than one `topologySpreadConstraint`, those constraints are ANDed: The kube-scheduler looks for a topology domain for the incoming Pod that satisfies all the constraints.

Reviewer: the scheduler looks for nodes, not for topology domains. Nodes have topology domains.

Author: I think this sentence is ok. Although scheduler looks for nodes from implementation details perspective, logically the scheduler looks for a topology domain which might be node/zone/region. There are other similar statements in the doc like 'the scheduler placed this incoming Pod into "zoneA"'. We don't have to say 'the scheduler placed this incoming Pod into a node in "zoneA"'.

Reviewer: But constraints are for different topology domains. So saying that a "topology domain satisfies all constraints" doesn't make much sense.

The correct interpretation is the "topology domains of a Node satisfy all constraints"

  - `ScheduleAnyway` tells the scheduler to still schedule it while prioritizing nodes that minimize the skew.
- **labelSelector** is used to find matching Pods. Pods that match this label selector are counted to determine the number of Pods in their corresponding topology domain. See [Label Selectors](/docs/concepts/overview/working-with-objects/labels/#label-selectors) for more details.

> When multiple `topologySpreadConstraint` are defined in a Pod spec, the constraints are ANDed. The kube-scheduler will look for a topology domain for the incoming Pod that satisfies all the constraints.

Reviewer 2: style nit:
```suggestion
When a Pod defines more than one `topologySpreadConstraint`, those constraints are ANDed. The kube-scheduler looks for a topology domain for the incoming Pod that satisfies all the constraints.
```

## Known Limitations

> - There's no guarantee that the constraints remain satisfied when Pods are removed. For example, scaling down a Deployment may result in imbalanced Pods distribution.

Reviewer 2: Totally optional: also mention descheduling and / or https://github.com/kubernetes-sigs/descheduler in the What's Next section.
